File: Leetcode/Javascript/4. Median of Two Sorted Arrays.js
var findMedianSortedArrays = function(nums1, nums2) {
    var size_1 = nums1.length;
    var size_2 = nums2.length;
    var sum = size_1+size_2;
    if(sum & 1) return help(nums1, 0, size_1-1, nums2, 0, size_2-1, (sum >> 1)+1);
    var first = help(nums1, 0, size_1-1, nums2, 0, size_2-1, (sum >> 1));
    var second = help(nums1, 0, size_1-1, nums2, 0, size_2-1, (sum >> 1)+1);
    console.log(first);
    console.log(second);
    return (first+second)/2;
};
function help(nums_1, start_1, end_1, nums_2, start_2, end_2, k) {
    var part_A = end_1-start_1+1;
    var part_B = end_2-start_2+1;
    //--------------------------//
    if(part_B < part_A) return help(nums_2, start_2, end_2, nums_1, start_1, end_1, k) ;
    if(!part_A) return nums_2[start_2+k-1];
    if(k === 1) return Math.min(nums_2[start_2], nums_1[start_1]);
    //-----------------------------------------------------------//
    part_A = Math.min(part_A, (k >> 1));
    part_B = k-part_A;
    //----------------------------------------------------------//
    if(nums_1[start_1+part_A-1] > nums_2[start_2+part_B-1]) return help(nums_1, start_1, end_1, nums_2, start_2+part_B, end_2,k-part_B);
    else if(nums_1[start_1+part_A-1] < nums_2[start_2+part_B-1]) return help(nums_1, start_1+part_A, end_1, nums_2, start_2, end_2,k-part_A);
    else return nums_1[start_1+part_A-1];
}

File: Leetcode/Three/4. Median of Two Sorted Arrays.java
public class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int size1 = nums1.length;
        int size2 = nums2.length;
        if(((size1+size2) & 1) == 1) return help(nums1, 0, size1-1, nums2, 0, size2-1, ((size1+size2)>>1)+1);
        else {
            int one = help(nums1, 0, size1-1, nums2, 0, size2-1, ((size1+size2)>>1));
            int two  =help(nums1, 0, size1-1, nums2, 0, size2-1, ((size1+size2)>>1)+1);
            return (one+two)/2.0;
        }
    }
    public int help(int[] one, int i, int j, int[] two, int n, int m, int k) {
        if(m-n > j-i) return help(two, n, m, one, i, j, k);
        if(n > m) return one[i+k-1];
        if(k == 1) return Math.min(one[i], two[n]);
        //----------------------------------------//
        int sizeTwo = Math.min((k >> 1), m-n+1);//for two
        int sizeOne = k-sizeTwo;//for one
        if(two[n+sizeTwo-1] <= one[i+sizeOne-1]) {
            return help(one, i, j, two, n+sizeTwo, m, k-sizeTwo);
        }
        else {
            return help(one, i+sizeOne, j, two, n, m, k-sizeOne);
        }
    }
}
